### PR TITLE
Unify plugins hook dispatch

### PR DESCRIPTION
**Pull request by dyrock**
* Added`HttpHookState` class that manages all http hooks. This will be useful for later plugin coordination work.
* Changed `ProxySession` and `HttpSM` to use `HttpHookState` for hooks dispatch
* Minor changes to `APIHook`, `APIHooks`, and `FeatureAPIHooks`

This is phase one of the [Plugin Coordination project](https://solidwallofcode.github.io/plugin-coordination.en.html). The early phases will provide mostly debugging support.

### CODE REVIEW

**Comment by shinrich** [comment]
The debian build problem just looks like a bad git request.  Probably just need to retry.  But the ubuntu build failure is a core during the regression test.  The stack has the hook logic on it which is worrying.

File: proxy/ProxySession.cc [hunk 1]
```diff
@@ -107,34 +101,29 @@ ProxySession::state_api_callout(int event, void *data)
...
+        if (!schedule_event) { // Don't bother if there is already one
+          schedule_event = mutex->thread_holding->schedule_in(this, HRTIME_MSECONDS(10));
         }
-
-        this->api_current = this->api_current->next();
-        hook->invoke(eventmap[this->api_hookid], this);
         return 0;
       }
+
+      SCOPED_MUTEX_LOCK(lock, hook->m_cont->mutex, mutex->thread_holding);
```
Comment (SolidWallOfCode) [inline proxy/ProxySession.cc]: You changed the `MUTEX_TRY_LOCK` to `SCOPED_MUTEX_LOCK` - I don't think that's a good idea.

Comment (dyrock) [reply proxy/ProxySession.cc]: Fei did it :/

Comment (duke8253) [reply proxy/ProxySession.cc]: Just to guarantee the lock is held when calling invoke. :D

**Comment by dragon512** [comment]
[approve ci autest]

File: configure.ac [hunk 2]
```diff
@@ -891,12 +891,12 @@ case $host_os_def in
...
-    ])
-    AS_IF([test -d /opt/local/lib], [
-      TS_ADDTO(AM_LDFLAGS, [-L/opt/local/lib])
-    ])
+    #AS_IF([test -d /opt/local/include], [
+    #  TS_ADDTO(TS_INCLUDES, [-I/opt/local/include])
+    #])
+    #AS_IF([test -d /opt/local/lib], [
+    #  TS_ADDTO(AM_LDFLAGS, [-L/opt/local/lib])
+    #])
```
Comment (shinrich) [inline configure.ac]: Seems unrelated to the rest of the PR

Comment (dyrock) [reply configure.ac]: oops. That is for compiling on my mac. I'll remove it :/

**Comment by zwoop** [comment]
[approve ci]

**Comment by randall** [comment]
[approve ci autest]

**Comment by shinrich** [comment]
[approve ci clang-analyzer]

**Comment by shinrich** [comment]
[approve ci clang-analyzer]

**Comment by duke8253** [comment]
[approve ci autest]

**Comment by duke8253** [comment]
[approve ci clang-analyzer]

File: proxy/InkAPIInternal.h [hunk 3]
```diff
@@ -198,28 +203,18 @@ template <typename ID, int N>
...
   }
+  m_hooks_p = false;
 }
 
 template <typename ID, int N>
 void
 FeatureAPIHooks<ID, N>::append(ID id, INKContInternal *cont)
 {
-  if (likely(is_valid(id))) {
-    hooks_p = true;
```
Comment (zwoop) [inline proxy/InkAPIInternal.h]: Is there a reason to remove the likely hints? At worst, they are pretty harmless right?

**Comment by zwoop** [comment]
Also this looks like an invasive change, with unclear reasons as to why this I needed. We really want this into 9.0 this late in the game ?

**Comment by shinrich** [comment]
[approve ci autest]

**Comment by shinrich** [comment]
Up to the release manager whether to pull this into 9.  We have been running with this code change for several months now.

**Comment by SolidWallOfCode** [comment]
It's not so late in the game, the PR has been here for quite a long time. And yes, I definitely want this in ATS 9 because it's a precursor to a lot of other work I want to get done in the ATS 10 time frame. We've been running with this internal for well over 9 months (we started production testing before this PR was created back in Feb).

**Comment by zwoop** [comment]
May I point out that I was right in questioning the sanity of including this in 9.0.x? :-P Thanks to Susan for fixing this bug (#6074), but clearly this was not well tested.
